### ISSUE TITLE
ISS: ebreak "retires" in debug mode, causing comparison mismatch

### ISSUE THREAD

**Issue by silabs-hfegran** [openhwgroup/core-v-verif#1650]
### Context
ebreak is an instruction that is not considered to retire, as per privileged spec. In the case below, we have a situation with an ebreak in the program-buffer that causes the debug code to restart, which appears to be correct for RTL (ebreak is reported with trap and exception on rvfi_trap), while the iss compares the ebreak-instruction with the first instruction of the handler in RTL, causing a mismatch and a subsequent off-by-one instruction in comparisons. 

### Steps to Reproduce
Please provide:
1. Branch: `https://github.com/silabs-hfegran/core-v-verif/tree/dev_hf_dbg_test_rewrite` , hash: 93a1e1e32e30a534ff806dfc05cf1f5a15a74dc9
2. Command line: `make test TEST=debug_test2 USE_ISS=1 RNDSEED=0`
3. Imperas version:
![image](https://user-images.githubusercontent.com/75469886/219032652-8f2396a1-dd4e-4a0b-89d4-bcc3767bb43e.png)

4. Tail end of simulation trace log:

>Info   x10 00000010 -> 00000005
Info 'refRoot/cpu', 0x000000001a11090a(_debugger_restore_stack+22): 0131     c.addi  x2,x2,12
Info   x2 1a111784 -> 1a111790
Info 'refRoot/cpu', 0x000000001a11090c(_debugger_restore_stack+24): bafe     cm.pop  {x1,x8-x9,x18-x27},112
Info   x1 1a110a2c -> 000002dc
Info   x2 1a111790 -> 1a111800
Info   x8 00000002 -> 00000000
Info   x9 00000002 -> 00000005
Info   x18 00000003 -> 00000000
Info   x19 00000003 -> 00000000
Info   x20 00000001 -> 00000000
Info   x21 1a110a76 -> 00000000
Info   x27 1a1108e8 -> 1a11095e
Info 'refRoot/cpu', 0x000000001a11090e(_debugger_restore_stack+26): 7b202173 csrr    x2,dscratch0
Info   x2 1a111800 -> 003fff50
Info 'refRoot/cpu', 0x000000001a110912(_debugger_restore_stack+2a): 7b341473 csrrw   x8,dscratch1,x8
Info   x8 00000000 -> 1a110a76
Info   dscratch1 1a110a76 -> 00000000
Info 'refRoot/cpu', 0x000000001a110916(_debugger_restore_stack+2e): 00801a63 bne     x0,x8,1a11092a
Info 'refRoot/cpu', 0x000000001a11092a(_debugger_restore_stack+42): 00040067 jr      x8
Info 'refRoot/cpu', 0x000000001a110a76(e1): 7b301af3 csrrw   x21,dscratch1,x0
Info 'refRoot/cpu', 0x000000001a110a7a(e1+4): 00100073 ebreak
Error (IDV) PC mismatch (HartId:0, PC:0x1a110a7a e1+4):
Info (IDV)  0>
Info (IDV)   . dut:0x1a110800 _debugger_start+0
Info (IDV)   . ref:0x1a110a7a e1+4
Error (IDV) Insn. bit pattern mismatch (HartId:0, PC:0x1a110a7a e1+4):
Info (IDV)  1>
Info (IDV)   . dut:7b211073 csrw    dscratch0,x2
Info (IDV)   . ref:00100073 ebreak
UVM_ERROR @ 240442.300 ns : rvvi-pkg.sv(94) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2api.state_compare @ 240442.000 ns: MISMATCH
Info 'refRoot/cpu', 0x000000001a110800(_debugger_start): 7b211073 csrw    dscratch0,x2
Error (IDV) PC mismatch (HartId:0, PC:0x1a110800 _debugger_start+0):
Info (IDV)  2>
Info (IDV)   . dut:0x1a110804 _debugger_start+4
Info (IDV)   . ref:0x1a110800 _debugger_start+0
Error (IDV) Insn. bit pattern mismatch (HartId:0, PC:0x1a110800 _debugger_start+0):
Info (IDV)  3>
Info (IDV)   . dut:00001117 auipc   x2,0x1
Info (IDV)   . ref:7b211073 csrw    dscratch0,x2
Error (IDV) GPR register value mismatch (HartId:0, PC:0x1a110800 _debugger_start+0):
Info (IDV)  4> GPR x2
Info (IDV)   . dut:0x1a111804 
Info (IDV)   . ref:0x003fff50 
UVM_ERROR @ 240451.300 ns : rvvi-pkg.sv(94) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2api.state_compare @ 240451.000 ns: MISMATCH
Info 'refRoot/cpu', 0x000000001a110804(_debugger_start+4): 00001117 auipc   x2,0x1
Info   x2 003fff50 -> 1a111804
Error (IDV) PC mismatch (HartId:0, PC:0x1a110804 _debugger_start+4):
Info (IDV)  5>
Info (IDV)   . dut:0x1a110808 _debugger_start+8
Info (IDV)   . ref:0x1a110804 _debugger_start+4
Error (IDV) Insn. bit pattern mismatch (HartId:0, PC:0x1a110804 _debugger_start+4):
Info (IDV)  6>
Info (IDV)   . dut:ffc10113 addi    x2,x2,-4
Info (IDV)   . ref:00001117 auipc   x2,0x1
Error (IDV) GPR register value mismatch (HartId:0, PC:0x1a110804 _debugger_start+4):
Info (IDV)  7> GPR x2
Info (IDV)   . dut:0x1a111800 
Info (IDV)   . ref:0x1a111804 
UVM_ERROR @ 240460.300 ns : rvvi-pkg.sv(94) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2api.state_compare @ 240460.000 ns: MISMATCH
Info 'refRoot/cpu', 0x000000001a110808(_debugger_start+8): ffc10113 addi    x2,x2,-4
Info   x2 1a111804 -> 1a111800
Error (IDV) PC mismatch (HartId:0, PC:0x1a110808 _debugger_start+8):
Info (IDV)  8>
Info (IDV)   . dut:0x1a11080c _debugger_start+12
Info (IDV)   . ref:0x1a110808 _debugger_start+8
Error (IDV) Insn. bit pattern mismatch (HartId:0, PC:0x1a110808 _debugger_start+8):
Info (IDV)  9>
Info (IDV)   . dut:b8fe     cm.push {x1,x8-x9,x18-x27},-112
Info (IDV)   . ref:ffc10113 addi    x2,x2,-4
Error (IDV) GPR register value mismatch (HartId:0, PC:0x1a110808 _debugger_start+8):
Info (IDV) 10> GPR x2
Info (IDV)   . dut:0x1a111790 
Info (IDV)   . ref:0x1a111800 
UVM_ERROR @ 240604.300 ns : rvvi-pkg.sv(94) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2api.state_compare @ 240604.000 ns: MISMATCH
Info 'refRoot/cpu', 0x000000001a11080c(_debugger_start+c): b8fe     cm.push {x1,x8-x9,x18-x27},-112
Info   x2 1a111800 -> 1a111790
Error (IDV) PC mismatch (HartId:0, PC:0x1a11080c _debugger_start+12):
Info (IDV) 11>
Info (IDV)   . dut:0x1a11080e _debugger_start+14
Info (IDV)   . ref:0x1a11080c _debugger_start+12
Error (IDV) Insn. bit pattern mismatch (HartId:0, PC:0x1a11080c _debugger_start+12):
Info (IDV) 12>
Info (IDV)   . dut:1151     c.addi  x2,x2,-12
Info (IDV)   . ref:b8fe     cm.push {x1,x8-x9,x18-x27},-112
Error (IDV) GPR register value mismatch (HartId:0, PC:0x1a11080c _debugger_start+12):
Info (IDV) 13> GPR x2
Info (IDV)   . dut:0x1a111784 
Info (IDV)   . ref:0x1a111790

**Comment by eroom1966** [comment]
Hi @silabs-hfegran (copying @Silabs-ArjanB )
This is a very strange one. What we have here is an ebreak executing whilst in debug mode.

Would we expect an ebreak in debug mode to take an exception, or to retire ?
From the debug spec it says
> When ebreak is executed (indicating the end of the Program Buffer code) the hart returns to its
> park loop

It does not say that it takes an exception, but what is being reported through the RVVI interface is the following, indicating an exception (EXC)
`UVM_INFO @ 240421.300 ns : rvvi-pkg.sv(57) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2log @ 240421.000 ns: EXC,0,22561,1a110a7a,"00100073 ebreak                 ",,,,,`

If it in fact did take an exception, then the next instruction to be executed, would be at the debug exception address, in your case 
`.dexc_address        = 0x1A111000,  // debug exception address`

However, the next instruction to execute by the RTL is at the debug vector address, ie the 'park loop'
`UVM_INFO @ 240442.300 ns : rvvi-pkg.sv(57) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2log @ 240442.000 ns: RET,0,22562,1a110800,"7b211073 csrw    dscratch0,x2   ",,,,,`

and from the configuration, the debug vector address is indeed this location
`.debug_address       = 0x1A110800,  // debug vector address`


The next question I would ask is, in this instance, should the execution of ebreak in debug mode increment the minstret ?
It would seem to me that it should

in short I think that the ebreak instruction appears to be modal
1. if ebreak executes in any mode other than debug, then it causes an exception, and does not increment minstret
2. if ebreak executes in debug mode, it retires, jumps to the debug vector address (and does increment minstret?)

I would really appreciate @Silabs-ArjanB comment on the above.

**Comment by Silabs-ArjanB** [comment]
Hi @eroom1966 

>> Would we expect an ebreak in debug mode to take an exception, or to retire ?

We expect it to take a 'trap'. I.e. in this case it takes a 'synchronous exception' with 'debug entry' as its target address (as opposed to the debug exception address). minstret shall not be incremented upon synchronous exceptions.

>     if ebreak executes in any mode other than debug, then it causes an exception, and does not increment minstret
>     if ebreak executes in debug mode, it retires, jumps to the debug vector address (and does increment minstret?)

I think it should be as follows:

- if ebreak executes in any mode other than debug, then it causes an exception, and does not increment minstret
- if ebreak executes in debug mode, then (also) it causes an exception (with target address equal to the park loop), and does not increment minstret

So in short: ebreak instructions never retire (as stated in the RISC-V Privileged spec) (if you think we are wrong here, please file a ticket with the Debug TG to get further clarification on this scenario, but we don't see any advantage in treating this scenario separately).

**Comment by eroom1966** [comment]
I think what makes me a little nervous was that we originally had this behavior, but there is a comment in the model when it changed in Sep 2020 for a bug we investigated from Steve Richmond and Oystein, and the comment is as follows

> Updates to debug mode to support the SiLabs testcase in the OpenHW Model
> correct dcsr.prv valid modes artifact register update, all non-ecall reported as ebreak

I am trying to find if we had a testcase exhibiting this change, this changed the code so that it did NOT raise a trap exception if the core was in debug mode and the cause was an ebreak

I will follow up

**Comment by eroom1966** [comment]
@silabs-hfegran 
If I revert the behavior to treat ebreak in debug mode as an exception - this test gets much further.
but it fails in this sequence

```
UVM_INFO @ 582010.300 ns : rvvi-pkg.sv(57) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2log @ 582010.000 ns: RET,0,54156,1a11095e,"7b200073 dret                   ",,,,,

UVM_INFO @ 582028.300 ns : rvvi-pkg.sv(57) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2log @ 582028.000 ns: EXC,0,54157,00001398,"4905     c.li    x18,1          ",x18=00000001,,,,

UVM_INFO @ 582043.300 ns : rvvi-pkg.sv(57) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2log @ 582043.000 ns: RET,0,54158,1a110800,"7b211073 csrw    dscratch0,x2   ",,,,CSR7b0(dcsr)=40008517 CSR7b1(dpc)=0000139a,

```
The instruction "c.li    x18,1" is reported as taking an exception, which is odd because of course it has written a value to x18, in which case it has retired and the exception occurs before executing the next instruction, in fact dpc has been loaded with the instruction at address 0x139a, so it seems to me that the flow should look as follows

```
UVM_INFO @ 582010.300 ns : rvvi-pkg.sv(57) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2log @ 582010.000 ns: RET,0,54156,1a11095e,"7b200073 dret                   ",,,,,

UVM_INFO @ 582028.300 ns : rvvi-pkg.sv(57) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2log @ 582028.000 ns: RET,0,54157,00001398,"4905     c.li    x18,1          ",x18=00000001,,,,

UVM_INFO @ xxxxxx.xxx ns : rvvi-pkg.sv(57) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2log @ xxxxxx.xxx ns: EXC,0,54157,0000139a,"4905    ???????          ",,,,,

UVM_INFO @ 582043.300 ns : rvvi-pkg.sv(57) reporter [] uvmt_cv32e40s_tb.imperas_dv.trace2log @ 582043.000 ns: RET,0,54158,1a110800,"7b211073 csrw    dscratch0,x2   ",,,,CSR7b0(dcsr)=40008517 CSR7b1(dpc)=0000139a,

```

I think we need to discuss this over a shared desktop.

**Comment by eroom1966** [comment]
Fixed by
https://github.com/openhwgroup/core-v-verif/pull/1659
And ImperasDV-OpenHW hash 490e1c0

**Comment by silabs-hfegran** [comment]
Resolved
